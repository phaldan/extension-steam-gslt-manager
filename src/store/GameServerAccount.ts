import { action, observable } from 'mobx';

export default class GameServerAccount {
  public steamid: string;
  public appid: number;

  @observable
  public expired: boolean = false;

  @observable
  public lastLogon?: Date;

  @observable
  public memo: string;

  @observable
  public token: string;

  @observable
  public locked: boolean = false;

  constructor({ steamid, appid, token, memo, lastLogon, expired }) {
    this.steamid = steamid;
    this.appid = appid;
    this.token = token;
    this.memo = memo;
    this.lastLogon = lastLogon ? new Date(lastLogon) : undefined;
    this.expired = expired;
  }

  @action
  public toggleLock(locked: boolean) {
    this.locked = locked;
  }

  @action
  public updateFromJson({ appid, token, memo, lastLogon, expired }) {
    this.appid = appid;
    this.token = token;
    this.memo = memo;
    this.lastLogon = lastLogon;
    this.expired = expired;
  }
}
