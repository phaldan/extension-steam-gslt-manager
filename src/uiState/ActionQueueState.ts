import { action, computed, observable } from 'mobx';
import ActionQueue from './ActionQueue';

export default class ActionQueueState {
  private removeDelay = 3000;

  @observable
  private queues: ActionQueue[] = [];

  @action
  public create(description: string, data: string[]): ActionQueue {
    const entry = new ActionQueue(this, this.removeDelay, description, data);
    this.queues.push(entry);
    return entry;
  }

  @computed
  public get running(): ActionQueue[] {
    return this.queues;
  }

  @action
  public async remove(id: string) {
    this.queues = this.queues.filter((q) => q.id !== id);
  }

  public setRemoveDelay(removeDelay: number): void {
    this.removeDelay = removeDelay;
  }
}
