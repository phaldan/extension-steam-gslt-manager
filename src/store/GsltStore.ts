import GameServerAccount from './GameServerAccount';

export default interface GsltStore {

  /**
   * Return current list of GameServerAccount.
   */
  readonly tokenAccounts: GameServerAccount[];

  readonly isLoggedIn: boolean;

  readonly isInitialized: boolean;

  /**
   * Load token accounts from steam.
   */
  loadAccounts(): Promise<void>;

  /**
   * Remove GSLT account.
   * @param account GameServerAccount instance
   */
  removeAccount(account: GameServerAccount): Promise<GameServerAccount>;

  /**
   * Remove GSLT accounts.
   * @param accounts List of GameServerAccount instances
   */
  removeAccounts(accounts: GameServerAccount[]): Array<Promise<GameServerAccount>>;

  /**
   * Create a new token for GSLT account.
   * @param account GameServerAccount instance
   */
  regenerateToken(account: GameServerAccount): Promise<GameServerAccount>;

  /**
   * Create a new token for GSLT accounts.
   * @param accounts List of GameServerAccount instances.
   */
  regenerateTokens(accounts: GameServerAccount[]): Array<Promise<GameServerAccount>>;

  /**
   * Change memo of GSLT account.
   * @param account GameServerAccount instance
   * @param memo Description for a GSLT
   */
  updateMemo(account: GameServerAccount, memo: string): Promise<GameServerAccount>;

  /**
   * Create multiple GSLT accounts.
   * @param amount Number of GSLT to create
   * @param appid Steam AppID
   * @param memo Description for a GSLT
   */
  createAccounts(amount: number, appid: string, memo: string): Array<Promise<void>>;
}
