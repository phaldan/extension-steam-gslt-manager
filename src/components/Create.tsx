import React from 'react';
import { ControlLabel, FormControl, FormGroup } from 'react-bootstrap/';
import GsltStore from '../store/GsltStore';
import ActionQueueState from '../uiState/ActionQueueState';
import ActionModal from './ActionModal';

interface CreateProps {
  store: GsltStore;
  text?: string;
  actions: ActionQueueState;
}

export default class Create extends React.Component<CreateProps> {
  public static defaultProps: Partial<CreateProps> = {
    text: 'New',
  };

  private amount: number = 1;
  private appid: string = '';
  private memo: string = '';

  public render() {
    const { text } = this.props;
    return (
      <ActionModal
        buttonIcon="plus"
        buttonText={text}
        title="Create GSLT"
        action="Create"
        onAction={this.save}
        onOpen={this.reset}
      >
        <FormGroup>
          <ControlLabel>
            Amount
          </ControlLabel>
          <FormControl
            type="number"
            defaultValue="1"
            onChange={this.setAmount}
            className="js-amount"
          />
        </FormGroup>
        <FormGroup>
          <ControlLabel>
            Game
          </ControlLabel>
          <FormControl
            type="text"
            placeholder="App-ID like 440 for TF2, 730 for CS:GO"
            onChange={this.setAppId}
            className="js-appid"
          />
        </FormGroup>
        <FormGroup>
          <ControlLabel>
            Memo
          </ControlLabel>
          <FormControl
            componentClass="textarea"
            placeholder="Optional short description"
            rows="10"
            maxLength="256"
            onChange={this.setMemo}
            className="js-memo"
          />
        </FormGroup>
      </ActionModal>
    );
  }

  private reset = () => {
    this.amount = 1;
    this.appid = '';
    this.memo = '';
  }

  private save = async () => {
    const { store, actions } = this.props;
    const amount = Number(this.amount);
    const list = Array.from(Array(Number(amount)), (_, i) => i.toString());
    const queue = actions.create(`Create ${amount} GSLTs:`, list);
    const promises = store.createAccounts(this.amount, this.appid, this.memo);
    const result = promises.map(async (promise, i: number) => {
      await promise;
      queue.progress(i.toString());
    });
    return Promise.all(result);
  }

  private setAmount = (event) => {
    this.amount = event.target.value;
  }

  private setAppId = (event) => {
    this.appid = event.target.value;
  }

  private setMemo = (event) => {
    this.memo = event.target.value;
  }
}
