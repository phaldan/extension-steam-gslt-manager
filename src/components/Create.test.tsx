import { shallow } from 'enzyme';
import * as React from 'react';
import ActionQueueState from 'uiState/ActionQueueState';
import GameServerAccount from 'store/GameServerAccount';
import GsltStore from 'store/GsltStore';
import Create from './Create';

function createComponent(store) {
    return shallow(<Create store={store} actions={new ActionQueueState()} />);
}

function expectComponent(target, method, calls) {
    expect(method).toHaveBeenCalledTimes(calls);
    expect(target).toMatchSnapshot();
}

describe('<Create>', () => {
    let store;
    let target;
    let createMock;
    let callback;

    function expectSuccesfullCreate(amount, appid, memo) {
        expectComponent(target.update(), createMock, 1);
        expect(createMock).toHaveBeenCalledWith(amount, appid, memo);
    }

    beforeEach(() => {
        store = new class implements GsltStore {
            loadAccounts(): Promise<void> {
                throw new Error("Method not implemented.");
            }
            get tokenAccounts(): GameServerAccount[] {
                throw new Error("Method not implemented.");
            }
            removeAccount(account: GameServerAccount): Promise<GameServerAccount> {
                throw new Error("Method not implemented.");
            }
            removeAccounts(accounts: GameServerAccount[]): Promise<GameServerAccount>[] {
                throw new Error("Method not implemented.");
            }
            regenerateToken(account: GameServerAccount): Promise<GameServerAccount> {
                throw new Error("Method not implemented.");
            }
            regenerateTokens(accounts: GameServerAccount[]): Promise<GameServerAccount>[] {
                throw new Error("Method not implemented.");
            }
            updateMemo(account: GameServerAccount, memo: string): Promise<GameServerAccount> {
                throw new Error("Method not implemented.");
            }
            createAccounts(amount: number, appid: string, memo: string): Promise<void>[] {
                createMock(amount, appid, memo);
                return [Promise.resolve(null)];
            }
            get isLoggedIn(): boolean {
                throw new Error("Method not implemented.");
            }
            get isInitialized(): boolean {
                throw new Error("Method not implemented.");
            }
        };
        createMock = jest.fn();
        target = createComponent(store);
    });

    test('fail to create account', () =>  {
        expectComponent(target, createMock, 0);
    });

    test('succesful create account', () =>  {
        target.find('ActionModal').simulate('action');
        expectSuccesfullCreate(1, '', '');
    });

    test('succesful change and create account', () =>  {
        target.find('.js-amount').simulate('change', {target: {value: 42}});
        target.find('.js-appid').simulate('change', {target: {value: '740'}});
        target.find('.js-memo').simulate('change', {target: {value: 'CSGO'}});
        target.find('ActionModal').simulate('action');
        expectSuccesfullCreate(42, '740', 'CSGO');
    });

    test('successfull reset of form data', () => {
        target.find('.js-amount').simulate('change', {target: {value: 42}});
        target.find('.js-appid').simulate('change', {target: {value: '740'}});
        target.find('.js-memo').simulate('change', {target: {value: 'CSGO'}});
        target.find('ActionModal').simulate('open');
        target.find('ActionModal').simulate('action');
        expectSuccesfullCreate(1, '', '');
    });
});
