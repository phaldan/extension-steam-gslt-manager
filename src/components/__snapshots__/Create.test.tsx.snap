// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Create> fail to create account 1`] = `
<ActionModal
  action="Create"
  buttonIcon="plus"
  buttonStyle="success"
  buttonText="New"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  onOpen={[Function]}
  title="Create GSLT"
>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Amount
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-amount"
      componentClass="input"
      defaultValue="1"
      onChange={[Function]}
      type="number"
    />
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Game
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-appid"
      componentClass="input"
      onChange={[Function]}
      placeholder="App-ID like 440 for TF2, 730 for CS:GO"
      type="text"
    />
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Memo
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-memo"
      componentClass="textarea"
      maxLength="256"
      onChange={[Function]}
      placeholder="Optional short description"
      rows="10"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Create> succesful change and create account 1`] = `
<ActionModal
  action="Create"
  buttonIcon="plus"
  buttonStyle="success"
  buttonText="New"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  onOpen={[Function]}
  title="Create GSLT"
>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Amount
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-amount"
      componentClass="input"
      defaultValue="1"
      onChange={[Function]}
      type="number"
    />
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Game
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-appid"
      componentClass="input"
      onChange={[Function]}
      placeholder="App-ID like 440 for TF2, 730 for CS:GO"
      type="text"
    />
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Memo
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-memo"
      componentClass="textarea"
      maxLength="256"
      onChange={[Function]}
      placeholder="Optional short description"
      rows="10"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Create> succesful create account 1`] = `
<ActionModal
  action="Create"
  buttonIcon="plus"
  buttonStyle="success"
  buttonText="New"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  onOpen={[Function]}
  title="Create GSLT"
>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Amount
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-amount"
      componentClass="input"
      defaultValue="1"
      onChange={[Function]}
      type="number"
    />
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Game
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-appid"
      componentClass="input"
      onChange={[Function]}
      placeholder="App-ID like 440 for TF2, 730 for CS:GO"
      type="text"
    />
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Memo
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-memo"
      componentClass="textarea"
      maxLength="256"
      onChange={[Function]}
      placeholder="Optional short description"
      rows="10"
    />
  </FormGroup>
</ActionModal>
`;

exports[`<Create> successfull reset of form data 1`] = `
<ActionModal
  action="Create"
  buttonIcon="plus"
  buttonStyle="success"
  buttonText="New"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  onOpen={[Function]}
  title="Create GSLT"
>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Amount
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-amount"
      componentClass="input"
      defaultValue="1"
      onChange={[Function]}
      type="number"
    />
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Game
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-appid"
      componentClass="input"
      onChange={[Function]}
      placeholder="App-ID like 440 for TF2, 730 for CS:GO"
      type="text"
    />
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Memo
    </ControlLabel>
    <FormControl
      bsClass="form-control"
      className="js-memo"
      componentClass="textarea"
      maxLength="256"
      onChange={[Function]}
      placeholder="Optional short description"
      rows="10"
    />
  </FormGroup>
</ActionModal>
`;
